domain/machine/service/service_test.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+)
+
+type serviceSuite struct {
+	testing.IsolationSuite
+	state *MockState
+}
+
+var _ = gc.Suite(&serviceSuite{})
+
+func (s *serviceSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+	s.state = NewMockState(ctrl)
+	return ctrl
+}
+
+func (s *serviceSuite) TestUpdateSuccess(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.state.EXPECT().UpsertMachine(gomock.Any(), "666").Return(nil)
+
+	err := NewService(s.state).Save(context.Background(), "666")
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *serviceSuite) TestUpdateError(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.state.EXPECT().UpsertMachine(gomock.Any(), "666").Return(errors.New("boom"))
+
+	err := NewService(s.state).Save(context.Background(), "666")
+	c.Assert(err, gc.ErrorMatches, `updating machine "666": boom`)
+}
+
+func (s *serviceSuite) TestDeleteSuccess(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.state.EXPECT().DeleteMachine(gomock.Any(), "666").Return(nil)
+
+	err := NewService(s.state).Delete(context.Background(), "666")
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *serviceSuite) TestDeleteError(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.state.EXPECT().DeleteMachine(gomock.Any(), "666").Return(errors.New("boom"))
+
+	err := NewService(s.state).Delete(context.Background(), "666")
+	c.Assert(err, gc.ErrorMatches, `deleting machine "666": boom`)
```

Comment:
Same comment as above. Better to check the error type and not the string. By checking the string we are saying this exact string is our interface.

Reply:
We typically use regexps in the `gc.ErrorMatches` just to check the core message we care about; the fluff around it is typically unimportant. Here the message was so simple a regexp was not needed. However, I'll make these requested changes here but we will have 10000000000000 other places to fix if we want to be consistent.

Reply:
Also, our test have very deliberately tested (part of) the error string as that forms part of the UX experience and we've caught various usability issues doing this.